unison-sync/src/Unison/Sync/Types.hs

```diff
@@ -0,0 +1,515 @@
+      [ "repo_name" .= name,
+        "path" .= Text.intercalate "." segments
+      ]
+
+instance FromJSON RepoPath where
+  parseJSON = Aeson.withObject "RepoPath" $ \obj -> do
+    repoName <- obj .: "repo_name"
+    pathSegments <- Text.splitOn "." <$> (obj .: "path")
+    pure RepoPath {..}
+
+newtype GetCausalHashByPathRequest = GetCausalHashByPathRequest
+  { repoPath :: RepoPath
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON GetCausalHashByPathRequest where
+  toJSON (GetCausalHashByPathRequest repoPath) =
+    object
+      [ "repo_path" .= repoPath
+      ]
+
+instance FromJSON GetCausalHashByPathRequest where
+  parseJSON = Aeson.withObject "GetCausalHashByPathRequest" $ \obj -> do
+    repoPath <- obj .: "repo_path"
+    pure GetCausalHashByPathRequest {..}
+
+newtype GetCausalHashByPathResponse = GetCausalHashByPathResponse
+  { causalHash :: HashJWT
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON GetCausalHashByPathResponse where
+  toJSON (GetCausalHashByPathResponse hashJWT) =
+    object
+      [ "causal_hash" .= hashJWT
+      ]
+
+instance FromJSON GetCausalHashByPathResponse where
+  parseJSON = Aeson.withObject "GetCausalHashByPathResponse" $ \obj -> do
+    causalHash <- obj .: "causal_hash"
+    pure GetCausalHashByPathResponse {..}
+
+data DownloadEntitiesRequest = DownloadEntitiesRequest
+  { repoName :: RepoName,
+    hashes :: NESet HashJWT
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON DownloadEntitiesRequest where
+  toJSON (DownloadEntitiesRequest repoName hashes) =
+    object
+      [ "repo_name" .= repoName,
+        "hashes" .= hashes
+      ]
+
+instance FromJSON DownloadEntitiesRequest where
+  parseJSON = Aeson.withObject "DownloadEntitiesRequest" $ \obj -> do
+    repoName <- obj .: "repo_name"
+    hashes <- obj .: "hashes"
+    pure DownloadEntitiesRequest {..}
+
+data DownloadEntitiesResponse = DownloadEntitiesResponse
+  { entities :: Map Hash (Entity HashJWT Hash Text)
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON DownloadEntitiesResponse where
+  toJSON (DownloadEntitiesResponse entities) =
+    object
+      [ "entities" .= entities
+      ]
+
+data UpdatePathRequest = UpdatePathRequest
+  { path :: RepoPath,
+    expectedHash :: Maybe TypedHash, -- Nothing requires empty history at destination
+    newHash :: TypedHash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON UpdatePathRequest where
+  toJSON (UpdatePathRequest path expectedHash newHash) =
+    object
+      [ "path" .= path,
+        "expected_hash" .= expectedHash,
+        "new_hash" .= newHash
+      ]
+
+instance FromJSON UpdatePathRequest where
+  parseJSON = Aeson.withObject "UpdatePathRequest" $ \obj -> do
+    path <- obj .: "path"
+    expectedHash <- obj .: "expected_hash"
+    newHash <- obj .: "new_hash"
+    pure UpdatePathRequest {..}
+
+-- | Not used in the servant API, but is a useful return type for clients to use.
+data UpdatePathResponse
+  = OutOfDate OutOfDateHash
```

Comment:
Could you rename `OutOfDateHash` (implies that the provided hash was at some point not out of date) to `CurrentHashMismatch` or something that just implies that it's wrong.

Reply:
And then update the dependency hash in Enlil 😅